Dsl/operator intergration (#136)
exasim-project/NeoFOAM: doc/finiteVolume/cellCentred/DSL.rst

The NeoFOAM DSL is designed as drop in replacement for OpenFOAM DSL and the adoption should be possible with minimal effort and the same equation from above should read:

.. code-block:: cpp

    dsl::EqnSystem<NeoFOAM::scalar> UEqn
    (
        fvcc::impOp::ddt(U)
        + fvcc::impOp::div(phi, U)
        - fvcc::impOp::laplacian(nu, U)
    )

    solve(UEqn == -fvcc::expOp::grad(p));

The majority of the work is done in the solve step: assemble the system and solve the system. After the system is solved or assembled, it allows needs to be provide access to the linear system for the SIMPLE and PISO algorithms.

EqnSystem
---------

The `EqnSystem` template class in NeoFOAM represents holds, manage, builds and solves the DSL and is the core part that orchestrates the following questions:

    - How to discretize the eqnterms?
        - In OpenFOAM this information is provided in **fvSchemes**
    - How to integrate the system in time?
        - In OpenFOAM this information is provided in **fvSchemes**
    - How to solve the system?
        - In OpenFOAM this information is provided in **fvSolution**

> The main difference between OpenFOAM and NeoFOAM is that the DSL in NeoFOAM is evaluated lazily. Therefore, the evaluation is no longer tied to a sparse matrix enabling other numerical integration strategies (RK method) or even substepping can be integrated inside the equation.

Reviewer: ```suggestion
The main difference between NeoFOAM and OpenFOAM is that the DSL is evaluated lazily. Therefore, the evaluation is not tied to the construction and solution of a sparse matrix, enabling other numerical integration strategies (e.g. RK methods or even substepping within an the equation).
```

> To evaluate the terms lazily, `EqnSystem` stores 3 vectors:

Reviewer: ```suggestion
For lazy evaluation, the `EqnSystem` stores 3 vectors:
```

.. mermaid::

    classDiagram
        class EqnTerm {

            +explicitOperation(...)
            +implicitOperation(...)
        }
        class DivEqnTerm {
            +explicitOperation(...)
            +implicitOperation(...)
        }
        class TemporalEqnTerm {
            +explicitOperation(...)
            +implicitOperation(...)
        }
        class Others["..."] {
            +explicitOperation(...)
            +implicitOperation(...)
        }
        class EqnSystem {
            +temporalTerms_: vector~EqnTerm~
            +implicitTerms_: vector~EqnTerm~
            +explicitTerms_: vector~EqnTerm~
        }
        EqnTerm <|-- DivEqnTerm
        EqnTerm <|-- TemporalEqnTerm
        EqnTerm <|-- Others
        EqnSystem <|-- EqnTerm

> So, the general assumption is that a EqnSystem consists of multiple EqnTerms that are either explicit, implicit or temporal. Consequently, plus, minus or scaling with a field needs to be handled by the EqnTerm.

Reviewer: ```suggestion
Thus, an `EqnSystem` consists of multiple `EqnTerms` which are either explicit, implicit, or temporal. Consequently, plus, minus, and scaling with a field needs to be handled by the `EqnTerm`.
```

EqnTerm
-------

> `EqnTerm` represents a term in an equation. It is a template class that can be instantiated with different value types. An `EqnTerm` can be explicit, implicit or temporal and needs be scalable by a scalar value or a field. `EqnTerm` is implemented as Type Erasure (more details `[1] <https://medium.com/@gealleh/type-erasure-idiom-in-c-0d1cb4f61cf0>`_ `[2] <https://www.youtube.com/watch?v=4eeESJQk-mw>`_ `[3] <https://www.youtube.com/watch?v=qn6OqefuH08>`_). So, the `EqnTerm` class provides a common interface for classes without inheritance. Consequently, the classes only needs to implement the interface and can be used in the DSL as shown in the example:

Reviewer: ```suggestion
The template `EqnTerm` represents a term in an equation and can be instantiated with different value types. An `EqnTerm` is either explicit, implicit or temporal, and needs to be scalable by a scalar value or a further field. The `EqnTerm` implementation used Type Erasure (more details `[1] <https://medium.com/@gealleh/type-erasure-idiom-in-c-0d1cb4f61cf0>`_ `[2] <https://www.youtube.com/watch?v=4eeESJQk-mw>`_ `[3] <https://www.youtube.com/watch?v=qn6OqefuH08>`_) to achieve polymorphism without inheritance. Consequently, the class needs only to implement the interface which is used in the DSL and which is shown in the below example:
```
just double get the grammar of that last sentence - still sounds a bit weird to me.

Example:
    .. code-block:: cpp

        NeoFOAM::DSL::EqnTerm<NeoFOAM::scalar> divTerm =

>             Divergence(NeoFOAM::DSL::EqnTerm<NeoFOAM::scalar>::Type::Explicit, exec, ...);

Reviewer: Just as a comment, I think we should create some 'typedefs' to make these guys more palatable. :)

Author: So we can reduce NeoFOAM::DSL::EqnTerm<NeoFOAM::scalar>::Type::Explicit to 
e.g.  dsl::Explicit
But this is a very good point

        NeoFOAM::DSL::EqnTerm<NeoFOAM::scalar> ddtTerm =
            TimeTerm(NeoFOAM::DSL::EqnTerm<NeoFOAM::scalar>::Type::Temporal, exec, ..);

> To fit the specification of the EqnSystem (storage in a vector), the EqnTerm needs to be able to be scaled:

Reviewer: I am not sure what is being said here - the 'storage in a vector' comment is unclear to me.

Author: That attempt should more accurately describe: To fit the specification of the EqnSystem 

EqnTerm is stored in the 3 vectors 
But it is not very clear

.. code-block:: cpp

        NeoFOAM::Field<NeoFOAM::scalar> scalingField(exec, nCells, 2.0);
        auto sF = scalingField.span();

        dsl::EqnTerm<NeoFOAM::scalar> customTerm =
            CustomTerm(dsl::EqnTerm<NeoFOAM::scalar>::Type::Explicit, exec, nCells, 1.0);

>         auto scaledTerm = 2.0 * customTerm; // 2.0 is the scaling factor
>         scaledTerm = -1.0 * customTerm; // -1.0 is the scaling factor
>         scaledTerm = scalingField * customTerm; // scalingField is the scaling field
> 
>         // EqnTerm also supports a similar syntax as OpenFOAM
>         auto scaledTerm = (scale + scale + scale + scale) * customTerm;

Reviewer: ```suggestion
        auto constantScaledTerm = 2.0 * customTerm; // A constant scaling factor of 2 for the term.
        auto fieldScaledTerm = scalingField * customTerm; // scalingField is used to scale the term.

        // EqnTerm also supports a similar syntax as OpenFOAM
        auto multiScaledTerm = (scale + scale + scale + scale) * customTerm;
```

        // EqnTerm also supports the use of a lambda as scaling function to reduce the number of temporaries generated

>         scaledTerm =

Reviewer: ```suggestion
        auto lambdaScaledTerm =
```

            (KOKKOS_LAMBDA(const NeoFOAM::size_t i) { return sF[i] + sF[i] + sF[i]  + sF[i]; }) * customTerm;

> The simplest approach to add a custom EqnTerm, is to derive a class from the EqnTermMixin and implement the required functions. A class that want to use the EqnTerm interface needs to provide the following functions:

Reviewer: ```suggestion
To add a user-defined `EqnTerm`, a new derived class must be created, inheriting from `EqnTermMixin`, 
 and provide the definitions of the below virtual functions that are required for the `EquTerm` interface: 
```

            termEvaluated = true;
        }

        std::string display() const { return "Laplacian"; }

        void explicitOperation(NeoFOAM::Field<NeoFOAM::scalar>& source)
        {
            NeoFOAM::scalar setValue = value;
            // scaleField is defined in EqnTermMixin
            // and accounts for the scaling of the terms
            // and considers scaling by fields and scalars
            auto scale = scaleField();
            auto sourceField = source.span();
            NeoFOAM::parallelFor(
                source.exec(),
                {0, source.size()},
                KOKKOS_LAMBDA(const size_t i) { sourceField[i] += scale[i] * setValue; }
            );
        }

        // other helper functions
        dsl::EqnTerm<NeoFOAM::scalar>::Type getType() const { return termType_; }

        const NeoFOAM::Executor& exec() const { return exec_; }

        std::size_t nCells() const { return nCells_; }

        fvcc::VolumeField<NeoFOAM::scalar>* volumeField() { return nullptr; }

        dsl::EqnTerm<NeoFOAM::scalar>::Type termType_;

        const NeoFOAM::Executor exec_;
        std::size_t nCells_;
        NeoFOAM::scalar value = 1.0;
    };

> The required scaling of the term is handle by the `scaleField` function that is provided by the EqnTermMixin. The `scaleField` function returns 'ScalingField' class that considers scaling by fields and scalars.

Reviewer: ```suggestion
The required scaling of the term is handle by the `scaleField` function, provided by `EqnTermMixin`. The `scaleField` function returns the 'ScalingField' class that is used to scale by fields and scalars.
```
